src/commons/crypto/signing/dispatch/krillsigner.rs

```diff
@@ -141,3 +186,512 @@ impl KrillSigner {
             .map_err(crypto::Error::signing)
     }
 }
+
+#[cfg(feature = "hsm")]
+impl KrillSigner {
+    fn build_signers(
+        signer_builder: SignerBuilderFn,
+        work_dir: &Path,
+        mapper: Arc<SignerMapper>,
+        configs: &[SignerConfig],
+    ) -> KrillResult<Vec<SignerProvider>> {
+        // For backward compatibility, no configured signers is the same as a single OpenSSL signer that will provide
+        // all signer functionality. To support running the Krill test suite against SoftHSM and PyKMIP we modify this
+        // behaviour slightly when special Krill Rust feature testing related flags are enabled.
+        let mut configs = configs.to_vec();
+        if configs.is_empty() {
+            configs.push(Self::get_default_signer_config()?);
+        }
+
+        // If there is only a single signer defined in the configuration file, don't require the user to specify it as
+        // the default signer. Also, for an OpenSSL signer, use it for all signing functions unless the operator
+        // explicitly said otherwise.
+        if configs.len() == 1 {
+            let config = &mut configs[0];
+
+            config.default.get_or_insert(true);
+
+            if matches!(config.signer_type, SignerType::OpenSsl(_)) {
+                config.oneoff.get_or_insert(true);
+                config.random.get_or_insert(true);
+            }
+        }
+
+        // One and only one signer should be the default. The default signer is used for operations that don't concern
```

Comment:
right, elsewhere we have been testing config in `Config::verify` - doing so here may keep this code simpler

Reply:
I put it here to try and keep multi-signer specific stuff out of "core" Krill. The rules are a bit more complicated than this comment alone, as demonstrated by the tests at the bottom of the file.

I'm not very happy about the number of things the test have to check, it makes me feel that the config syntax for multi-signers is wrong and should be revisited. E.g. instead of each signer defining if it is default, oneoff and/or random fallback signer, perhaps those should be top level settings like `auth_type`, e.g. `default_signer = ...` and `random_fallback_signer = `. The problem with the latter approach however is that you then have to name signers so that you can refer to them elsewhere in the config, and you separate something about the signer out from the signer itself. 

Ideas welcome!

Reply:
ok, let's keep it here for now